module_ctx.go

```diff
@@ -998,57 +998,73 @@ func (mctx *mutatorContext) Module() Module {
 	return mctx.module.logicModule
 }
 
-func (mctx *mutatorContext) AddDependency(module Module, tag DependencyTag, deps ...string) {
+func (mctx *mutatorContext) AddDependency(module Module, tag DependencyTag, deps ...string) []Module {
+	depInfos := make([]Module, 0, len(deps))
 	for _, dep := range deps {
 		modInfo := mctx.context.moduleInfo[module]
-		errs := mctx.context.addDependency(modInfo, tag, dep)
+		depInfo, errs := mctx.context.addDependency(modInfo, tag, dep)
 		if len(errs) > 0 {
 			mctx.errs = append(mctx.errs, errs...)
 		}
+		if depInfo != nil {
+			depInfos = append(depInfos, depInfo.logicModule)
```

Comment:
This is kind of odd behavior, the length of the return value is different than the length of the deps in the case of missing dependencies.  This seems likely to cause common panics for missing dependencies, for example:

```
dep := mctx.AddDependency("foo")[0] // panic here due to an index out of range error
transitive = append(transitive, dep.transitive...)
```

The alternative would be to return nils in the slice for missing dependencies, but that just delays the problem:
```
dep := mctx.AddDependency("foo")[0]
transitive = append(transitive, dep.transitive...) // panic here due to nil pointer dererference
```

Either way is going to need a bunch of error checking by anyone who uses the return value.  I think the second is a little easier to reason about though.

Reply:
It is slightly more convenient to iterate over a shorter or empty slice than over a slice with nils, but I agree that returning nils is less unexpected. Fixed to return a slice with possible nil entries (and mentioned that in the comments above).